pkgs/development/lisp-modules/default.nix

```diff
@@ -0,0 +1,299 @@
+
+      # Some libraries have multiple systems under one project, for
+      # example, cffi has cffi-grovel, cffi-toolchain etc.  By
+      # default, only the `pname` system is build.
+      #
+      # .asd's not listed in `systems` are removed in
+      # installPhase. This prevents asdf from referring to uncompiled
+      # systems on run time.
+      #
+      # Also useful when the pname is differrent than the system name,
+      # such as when using reverse domain naming.
+      systems ? [ pname ],
+
+      # The .asd files that this package provides
+      # TODO(kasper): remove
+      asds ? systems,
+
+      # Other args to mkDerivation
+      ...
+    } @ args:
+
+    stdenv.mkDerivation (rec {
+      inherit
+        pname version nativeLibs javaLibs lispLibs systems asds
+        pkg program flags faslExt
+      ;
+
+      # When src is null, we are building a lispWithPackages and only
+      # want to make use of the dependency environment variables
+      # generated by build-asdf-system
+      dontUnpack = src == null;
+
+      # Portable script to build the systems.
+      #
+      # `lisp` must evaluate this file then exit immediately. For
+      # example, SBCL's --script flag does just that.
+      #
+      # NOTE:
+      # Every other library worked fine with asdf:compile-system in
+      # buildScript.
+      #
+      # cl-syslog, for some reason, signals that CL-SYSLOG::VALID-SD-ID-P
+      # is undefined with compile-system, but works perfectly with
+      # load-system. Strange.
+
+      # TODO(kasper) portable quit
+      asdfFasl = buildAsdf { inherit asdf pkg program flags faslExt; };
+
+      buildScript = substituteAll {
+        src = ./builder.lisp;
+        asdf = "${asdfFasl}/asdf.${faslExt}";
+      };
+
+      preConfigure = ''
+        source ${./setup-hook.sh}
+        buildAsdfPath
+      '';
+
+      buildPhase = optionalString (src != null) ''
+        export CL_SOURCE_REGISTRY=$CL_SOURCE_REGISTRY:$src//
+        export ASDF_OUTPUT_TRANSLATIONS="$src:$(pwd):${storeDir}:${storeDir}"
+        ${pkg}/bin/${program} ${flags} < $buildScript
+      '';
+
+      # Copy compiled files to store
+      #
+      # Make sure to include '$' in regex to prevent skipping
+      # stuff like 'iolib.asdf.asd' for system 'iolib.asd'
+      #
+      # Same with '/': `local-time.asd` for system `cl-postgres+local-time.asd`
+      installPhase =
+        let
+          mkSystemsRegex = systems:
+            concatMapStringsSep "\\|" (replaceStrings ["." "+"] ["[.]" "[+]"]) systems;
+        in
+      ''
+        mkdir -pv $out
+        cp -r * $out
+
+        # Remove all .asd files except for those in `systems`.
+        find $out -name "*.asd" \
+        | grep -v "/\(${mkSystemsRegex systems}\)\.asd$" \
+        | xargs rm -fv || true
+      '';
+
+      dontPatchShebangs = true;
+
+      # Not sure if it's needed, but caused problems with SBCL
+      # save-lisp-and-die binaries in the past
+      dontStrip = true;
```

Comment:
we should verify if this is still true

Reply:
SBCL still appends (in the file concatenation sense) the «image» after the small runtime kernel, strip will still remove it, and I don't see either side changing the behaviour anytime soon.